kernel/timeout.c

```diff
@@ -46,11 +46,14 @@ static struct _timeout *next(struct _timeout *t)
 
 static void remove_timeout(struct _timeout *t)
 {
-	if (next(t) != NULL) {
-		next(t)->dticks += t->dticks;
-	}
+	if (t->node.next != NULL && t->node.prev != NULL) {
```

Comment:
I'd considered `clear` instead of `unlink` but that didn't seem as...clear.  If there's another wording that's better....

Alternatively, put the `node->next = NULL` assignment into `remove`, which would be safer anyway at a very minor overall cost.  An ability to reliably check whether a node is already in a list before adding it to a list could be very helpful.

Reply:
I generally support @pabigot idea, however I would like to add small modification:
To be consistent with the way how list is implemented in Zephyr, I suggest following Linux (https://elixir.bootlin.com/linux/latest/ident/list_del_init) and mark removed node by setting `node->next` (and `node->prev`) to `node`.

Reply:
Researching what @pizi-nordic proposed:  In Linux `list_del_init` removes the entry and initializes the entry as an empty list head.  `list_del` removes the entry and sets the `next` and `prev` pointers to poisoned values (dereference causes a page fault that can be diagnosed).  `list_del` is used about twice as often as `list_del_init` in Linux, but `list_del_init` is to be used [when the removed entry may later be added to another list](https://marc.info/?l=linux1394-devel&m=114328842931957).

Although the structure of a doubly-linked list is the same in Linux and Zephyr the pointer management is different in both operations supported and order-of-assignment for specific operations like add/prepend.  Zephyr uses distinct type names for list nodes and list heads and they aren't initialized the same way, unlike Linux which only uses `list_head`.   So I'm not convinced that resetting a removed `sys_dnode_t` to appear like an empty `sys_dlist_t` should be preferred over initializing a removed node to have null pointers, which is presumably what a Zephyr `sys_dnode_t` that's never been added to a list would have (if it isn't left uninitialized).

In either case, after reviewing this I do think it's important that both `next` and `prev` be assigned to self or a poison/null value when the entry is removed, not just `next` as I originally proposed.

Reply:
@andyross Do you mean you're OK with dropping this patch from this PR, or that you'd rather merge this change as-is rather than bike-shed it here?

Either is fine with me.  I will file a PR to fix dlist, because as you've found the existing code is problematic and it should be fixed more generally.  IMO there's too much "hack it now, we'll fix it properly later" in Zephyr, and later never comes.

Reply:
If the problem is fixed globally about 42 tests fail sanity check.  This after working around a bug in poll which double-removes events from a list (segfaults in many native_posix tests). Looks like several parts of Zephyr depend on the pointers to neighboring nodes remaining dereferenceable after removal from a list.

Probably best to merge your fix for now.